### ISSUE TITLE
Error in predict.randomForest when running Chapter 1 locally

### ISSUE THREAD

**Issue by kbajdzienko** [juliasilge/supervised-ML-case-studies-course#59]
First of all, this tutorial is amazing. Content, pace, level of detail. I love it.

I encountered one issue with random forest when following the code of Chapter 1 locally. I'm on tidymodels 0.1.2 and randomForest 4.6-14 running in Windows.

While I found a solution by mutating chr cols in `car_vars` to factor, I have no idea why the code that works on [netlify](https://supervised-ml-course.netlify.app/chapter1) did not work locally. 

Running `predict(fit_rf, car_train)` returned: 
```
Error in predict.randomForest(object = object$fit, newdata = new_data) : 
New factor levels not present in the training data
```

To reproduce:
```
install.packages(c("tidymodels","randomForest"))
library(tidymodels)
csv_url <- "https://raw.githubusercontent.com/juliasilge/supervised-ML-case-studies-course/master/data/cars2018.csv"
download.file(csv_url,"cars.csv")
cars <- readr::read_csv("cars.csv")

set.seed(1234)

car_vars <- cars %>%
  select(-model, -model_index)

car_split <- car_vars %>%
  initial_split(prop = 0.8,
                strata = aspiration)

car_train <- training(car_split)

rf_mod <- rand_forest() %>%
  set_mode("regression") %>%
  set_engine("randomForest")

fit_rf <- rf_mod %>%
  fit(log(mpg) ~ ., 
      data = car_train) 

results <- car_train %>%
  mutate(mpg = log(mpg)) %>%
  bind_cols(predict(fit_rf, car_train) %>%
              rename(.pred_rf = .pred))
```
What I noticed is that all levels in `str(fit_rf[["fit"]][["forest"]][["xlevels"]])` were numeric (contrary to the model stored in `data/c1_fit_rf.rds`. Maybe someone here could explain me why, since randomForest is new to me?

Solution to the error was to enforce factor class on chrs: `car_vars <- mutate(car_vars, across(where(is.character),as.factor))`

**Comment by juliasilge** [comment]
Hmmmmm, that seems quite mysterious to me; the randomForest package [hasn't had a release since 2018](https://cran.r-project.org/package=randomForest) so I'm sure it's something in how a tidymodels package has changed recently. My best guess would be the [categorical encoding changes we discussed here](https://www.tidyverse.org/blog/2020/07/parsnip-0-1-2/#predictor-encoding-consistency).

I'm going to leave this open so I can come back and update the course (including the pretrained model `.rds` file).

**Comment by juliasilge** [comment]
For the record, we do typically recommend using factors when building models, so you can "be the boss" of what factor levels are in your model, how to account for unseen factor levels, etc.

**Comment by RaymondBalise** [comment]
Hello Julia,
As always, this is *fabulous* pedagogy.  I am planning on using this when I teach ML over in a June.  I would be great if chapter 1, topic 4 could be tweaked to say:

```
# Create cars_vars after deselecting the 2 columns that are record IDs and ...
car_vars <- cars2018 %>%
    select(-model, -model_index) |>  
    # ... converting the character strings into categorical factors.
    mutate(across(where(is.character), as.factor)) 
```

I am thinking that solves the problem and mentioning the fact that these two things are IDs sets the stage for latter explaining `update_role()`.

What do you think?

**Comment by juliasilge** [comment]
Yes, some kind of change like this is needed. ~~With MRAN closing up shop, I will need to do some more general maintenance on this course, so I'll take care of this at that point.~~ Looks I already switched from MRAN to RSPM, but there still is some general updating that needs to happen.

Did you say you are wanting to use this for a course this coming June?

**Comment by RaymondBalise** [comment]
Yes.  I am teaching reproducible research this semester and ML over the summer.  As you can tell I am a fan.  If you would like to collaborate, I will find time to help in any way I can.

Ray

Raymond R. Balise, PhD
Associate  Professor
Department of Public Health Sciences, Biostatistics

University of Miami, Miller School of Medicine
1120 N.W. 14th Street
Don Soffer Clinical Research Center - Room 1061
Miami, Florida 33136

(305) 243-9694

To request help from the Center for HIV and Research in Mental Health (CHARM) please go here<https://redcap.miami.edu/surveys/?s=EEWDYECME7>.  To provide post-consult feedback go here<https://redcap.miami.edu/surveys/?s=EYMEPRMCJP&consultant=1>. Please remember to acknowledge CHARM in your papers if resources were used (P30MH116867).

To request help with SCAN360 please go here<https://redcap.miami.edu/surveys/?s=Y7DMMNHFKAEYDH3W>.

Confidentiality Notice: This e-mail transmission, and any documents, files or previous e-mail messages attached to it, may contain confidential information. If you are not the intended recipient, or a person responsible for delivering it to the intended recipient, you are hereby notified that any disclosure, copying, distribution or use of any of the information contained in or attached to this message is STRICTLY PROHIBITED. If you have received this transmission in error, please immediately notify us by reply e-mail or by telephone at (305) 243-9694  and destroy the original transmission and its attachments without reading them or saving them to disk.


From: Julia Silge ***@***.***>
Date: Friday, January 13, 2023 at 10:16 AM
To: juliasilge/supervised-ML-case-studies-course ***@***.***>
Cc: Balise, Raymond R ***@***.***>, Comment ***@***.***>
Subject: [EXTERNAL] Re: [juliasilge/supervised-ML-case-studies-course] Error in predict.randomForest when running Chapter 1 locally (#59)
CAUTION: This email originated from outside the organization. DO NOT CLICK ON LINKS or OPEN ATTACHMENTS unless you know and trust the sender.

Yes, some kind of change like this is needed. With MRAN closing up shop, I will need to do some more general maintenance on this course, so I'll take care of this at that point. Did you say you are wanting to use this for a course this coming June?
